pkg/controller/services.go

```diff
@@ -273,6 +274,70 @@ func (cont *AciController) updateServicesForNode(nodename string) {
 	cont.serviceEndPoints.UpdateServicesForNode(nodename)
 }
 
+// must have index lock
+func (cont *AciController) getActiveFabricPathDn(node string) string {
+	var fabricPathDn string
+	sz := len(cont.nodeOpflexDevice[node])
+	for i := range cont.nodeOpflexDevice[node] {
+		device := cont.nodeOpflexDevice[node][sz-1-i]
+		if device.GetAttrStr("state") == "connected" {
+			fabricPathDn = device.GetAttrStr("fabricPathDn")
+			break
+		}
+	}
+	return fabricPathDn
+}
+
+func (cont *AciController) deleteOldOpflexDevices() {
+	var nodeUpdates []string
+	cont.indexMutex.Lock()
+	for node, devices := range cont.nodeOpflexDevice {
+		fabricPathDn := cont.getActiveFabricPathDn(node)
+		if fabricPathDn != "" {
+			updated := false
+			for i, device := range devices {
+				if device.GetAttrStr("delete") == "true" && device.GetAttrStr("fabricPathDn") != fabricPathDn {
+					deleteTimeStr := device.GetAttrStr("deleteTime")
+					deleteTime, err := time.Parse(time.RFC3339, deleteTimeStr)
+					if err != nil {
+						cont.log.Error("Failed to parse opflex device delete time: ", err)
+						continue
+					}
+					now := time.Now()
+					diff := now.Sub(deleteTime)
+					if diff.Seconds() >= cont.config.DeviceDeleteTimeout {
+						devices = append(devices[:i], devices[i+1:]...)
```

Comment:
I am not sure I understand the logic here, aren't you iterating over "devices"? So, if L308 is true, it means you found a candidate device to delete? And if so, should you not make note of that device in a temporary list and set a flag "update" to True. L310 would not be required. Outside this for loop you check if the temporary list in not empty, and if so iterate through this temp list to delete the devices for this node (and if deleted add this node to the nodeUpdates list).